### PR TITLE
Move modlog writing to Room.prototype

### PR DESCRIPTION
**Pull request by bumbadadabum**
This allows chat plugins (such as the wifi giveaway plugin) to modnote things, which was previously quite a large limitation.

### CODE REVIEW

File: rooms.js [hunk 1]
```diff
@@ -906,6 +941,8 @@ let BattleRoom = (() => {
 	BattleRoom.prototype.expireTimer = null;
 	BattleRoom.prototype.active = false;
 
+	BattleRoom.prototype.modlog = fs.createWriteStream(path.resolve(__dirname, 'logs/modlog/modlog_battle.txt'), {flags:'a+'});
```
Comment (Slayer95) [inline rooms.js]: *shudders*

Comment (bumbadadabum) [inline rooms.js]: Right now it also creates the writestream by default, so I didn't see this as a problem.

**Comment by Slayer95** [comment]
These look like nice additions to ``Room``, but ``CommandContext`` is fine imo.

Whenever you __need__ to call ``logModCommand`` from ``room``, your aren't actually logging a command... And ``Room#globalModlog`` is a __terrible__ API...

What do you say about ``CommandContext#logModCommand`` = ``Room#modLog()`` and ``CommandContext#globalModlog`` -> ``Rooms.global.writeModlog`` ?

File: rooms.js (hunk 1 above)
Comment (Slayer95) [inline rooms.js]: Well, this can't be rewritten to ``class`` notation. Not that I care, but probably @Zarel does.

**Comment by bumbadadabum** [comment]
That makes sense, yeah. I'll change that. Also, the reason I had globalModlog in Room is because this way it doesn't require the room itself to be passed as an arg as well.

**Comment by Slayer95** [comment]
Yea well, with ``this.globalModlog`` you don't need to pass the room either.

A plugin writing to the global log seems like an extremely niche use case.

**Comment by bumbadadabum** [comment]
Fair enough.

**Comment by bumbadadabum** [comment]
Updated.

File: rooms.js [hunk 2]
```diff
@@ -306,6 +307,22 @@ let Room = (() => {
...
+	Room.prototype.sendModCommand = function (data) {
+		for (let i in this.users) {
+			let user = this.users[i];
+			// hardcoded for performance reasons (this is an inner loop)
+			if (user.isStaff || (this.auth && (this.auth[user.userid] || '+') !== '+')) {
+				user.sendTo(this, data);
+			}
+		}
+	};
+	Room.prototype.modLog = function (text) {
```
Comment (Zarel) [inline rooms.js]: Let's capitalize consistently: `modlog`

File: rooms.js [hunk 3]
```diff
@@ -432,6 +449,9 @@ let GlobalRoom = (() => {
 			() => this.periodicMatch(),
 			PERIODIC_MATCH_INTERVAL
 		);
+
+		// Create writestream for modlog
+		this.modlog = fs.createWriteStream(path.resolve(__dirname, 'logs/modlog/modlog_global.txt'), {flags:'a+'});
```
Comment (Zarel) [inline rooms.js]: You can use `.modlogStream` for the modlog writestream.

File: rooms.js (hunk 2 above)
Comment (Zarel) [inline rooms.js]: Also having two different functions `modlog` and `writemodlog` is kind of confusing - do we really need both?

File: rooms.js (hunk 1 above)
Comment (Zarel) [inline rooms.js]: I would probably just put it in the constructor.

Comment (Zarel) [inline rooms.js]: You could I guess set up `Room.globalModlog` and `Rooms.groupchatModlog`...

Comment (bumbadadabum) [inline rooms.js]: Wouldn't putting it in the constructor make a separate write stream for
every battleroom?

**Comment by bumbadadabum** [comment]
Ok did things.

File: rooms.js [hunk 4]
```diff
@@ -906,6 +926,8 @@ let BattleRoom = (() => {
 	BattleRoom.prototype.expireTimer = null;
 	BattleRoom.prototype.active = false;
 
+	BattleRoom.prototype.modlogStream = fs.createWriteStream(path.resolve(__dirname, 'logs/modlog/modlog_battle.txt'), {flags:'a+'});
```
Comment (Zarel) [inline rooms.js]: I saaaaid put this in the constructor :|

File: rooms.js [hunk 5]
```diff
@@ -1366,9 +1388,12 @@ let ChatRoom = (() => {
 			this.userList = this.getUserList();
 			this.reportJoinsQueue = [];
 		}
+
+		if (!this.isPersonal) this.modlogStream = fs.createWriteStream(path.resolve(__dirname, 'logs/modlog/modlog_' + roomid + '.txt'), {flags:'a+'});
 	}
 	ChatRoom.prototype = Object.create(Room.prototype);
 	ChatRoom.prototype.type = 'chat';
+	ChatRoom.prototype.modlogStream = fs.createWriteStream(path.resolve(__dirname, 'logs/modlog/modlog_groupchat.txt'), {flags:'a+'});
```
Comment (Zarel) [inline rooms.js]: Ditto

**Pull request merged**